src/backend/distributed/commands/multi_copy.c

```diff
@@ -462,6 +499,160 @@ CitusCopyFrom(CopyStmt *copyStatement, char *completionTag)
+	CopyState copyState = BeginCopyFrom(distributedRelation,
+										copyStatement->filename,
+										copyStatement->is_program,
+										copyStatement->attlist,
+										copyStatement->options);
+
+	CopyOutState copyOutState = (CopyOutState) palloc0(sizeof(CopyOutStateData));
+	copyOutState->binary = true;
+	copyOutState->fe_msgbuf = makeStringInfo();
+	copyOutState->rowcontext = executorTupleContext;
+
+	columnOutputFunctions = ColumnOutputFunctions(tupleDescriptor, copyOutState->binary);
+
+	/* we use a PG_TRY block to close connections on errors (e.g. in NextCopyFrom) */
+	PG_TRY();
+	{
+		uint64 shardMaxSizeInBytes = (int64) ShardMaxSize * 1024L;
+		uint64 copiedDataSizeInBytes = 0;
+		uint64 processedRowCount = 0;
+
+		/* set up callback to identify error line number */
+		ErrorContextCallback errorCallback;
+
+		errorCallback.callback = CopyFromErrorCallback;
+		errorCallback.arg = (void *) copyState;
+		errorCallback.previous = error_context_stack;
+		error_context_stack = &errorCallback;
+
+		while (true)
+		{
+			bool nextRowFound = false;
+			MemoryContext oldContext = NULL;
+			uint64 messageBufferSize = 0;
+
+			ResetPerTupleExprContext(executorState);
+
+			oldContext = MemoryContextSwitchTo(executorTupleContext);
+
+			/* parse a row from the input */
+			nextRowFound = NextCopyFrom(copyState, executorExpressionContext,
+										columnValues, columnNulls, NULL);
+
+			if (!nextRowFound)
+			{
+				MemoryContextSwitchTo(oldContext);
+				break;
+			}
+
+			CHECK_FOR_INTERRUPTS();
+
+			MemoryContextSwitchTo(oldContext);
+
+			/*
+			 * If copied data size is zero, this means either this is the first
+			 * line in the copy or we just filled the previous shard up to its
+			 * capacity. Either way, we need to create a new shard and
+			 * start copying new rows into it.
+			 */
+			if (copiedDataSizeInBytes == 0)
+			{
+				StartCopyToNewShard(copyOutState, shardConnections, relationId,
+									copyStatement);
+			}
+
+			/* replicate row to shard placements */
+			resetStringInfo(copyOutState->fe_msgbuf);
+			AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
+							  copyOutState, columnOutputFunctions);
+			SendCopyDataToAll(copyOutState->fe_msgbuf, shardConnections->connectionList);
+
+			messageBufferSize = copyOutState->fe_msgbuf->len;
+			copiedDataSizeInBytes = copiedDataSizeInBytes + messageBufferSize;
+
+			/*
+			 * If we filled up this shard to its capacity, finalize the copy
+			 * command by sending copy footers and calculating shard statistics.
+			 */
+			if (copiedDataSizeInBytes > shardMaxSizeInBytes)
+			{
+				FinalizeCopyToNewShard(copyOutState, shardConnections, relationId);
+				copiedDataSizeInBytes = 0;
+			}
+
+			processedRowCount += 1;
+		}
+
+		/* all lines have been copied, stop showing line number in errors */
+		error_context_stack = errorCallback.previous;
+
+		/* check for cancellation one last time before committing */
+		CHECK_FOR_INTERRUPTS();
```

Comment:
FinalizeCopyToNewShard might actually take a while. It might be better to check afterwards.